File: bubblesort.hs


--
bubblesort :: [Int] -> [Int]
bubblesort arr
    | (cnt == 0) || (length arr <= 2) = new
    | otherwise = (bubblesort (take ((length arr)-1) new)) ++ [(new !! ((length arr)-1))]
    where
        (cnt, new) = sort arr
-- 
sort :: [Int] -> (Int, [Int])
sort [] = (0, [])
sort (x:xs)
    | null xs       = (0, [x])
    | x > xs1   = (cnt +1, xs1 : arr)
    | otherwise     = (cnt, x : arr)
    where
        xs1 = xs !! 0
        (cnt, arr) = sort ((maximum (x:[(xs!!0)])) : (drop 1 xs))

File: bubblesort.py
def bubblesort(arr):
    cnt = 1
    while cnt > 0:
        cnt = 0
        # da immer i mit (i+1) verglichen wird nur bis len-1
        for i in range(0,len(arr)-1):   
            if arr[i] > arr[i+1]:
                tmp = arr[i]
                arr[i] = arr[i+1]
                arr[i+1] = tmp
                cnt += 1
    return arr

def main():
    res = bubblesort([3,2,1,6,5,5,5,5,9,0,9,8,1])
    print(res)

if __name__=="__main__": 
    main() 
